<a href="https://colab.research.google.com/github/fralfaro/MAT306/blob/main/docs/labs/lab_04.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# MAT306 - Laboratorio N°04


**Objetivo**: Aplicar técnicas intermedias y avanzadas de análisis de datos con pandas utilizando un caso real: el Índice de Libertad de Prensa. Este laboratorio incluye operaciones de limpieza, transformación, combinación de datos, y análisis exploratorio usando `merge`, `groupby`, `concat` y otras funciones fundamentales.




**Descripción del Dataset**

El presente conjunto de datos está orientado al análisis del **Índice de Libertad de Prensa**, una métrica internacional que evalúa el nivel de libertad del que gozan periodistas y medios de comunicación en distintos países. Este índice es recopilado anualmente por la organización **Reporteros sin Fronteras**.

La base de datos contempla observaciones por país y año, e incluye tanto el valor del índice como el ranking correspondiente. A menor puntaje en el índice, mayor nivel de libertad de prensa.

**Diccionario de variables**

| Variable     | Clase    | Descripción                                                                          |
| ------------ | -------- | ------------------------------------------------------------------------------------ |
| `codigo_iso` | carácter | Código ISO 3166-1 alfa-3 que representa a cada país.                                 |
| `pais`       | carácter | Nombre oficial del país.                                                             |
| `anio`       | entero   | Año en que se registró la medición del índice.                                       |
| `indice`     | numérico | Valor numérico del Índice de Libertad de Prensa (menor valor indica mayor libertad). |
| `ranking`    | entero   | Posición relativa del país en el ranking mundial de libertad de prensa.              |


**Fuente original y adaptación pedagógica**

* **Fuente original**: [Reporteros sin Fronteras](https://www.rsf-es.org/), recopilado y publicado a través del portal del [Banco Mundial](https://tcdata360.worldbank.org/indicators/h3f86901f?country=BRA&indicator=32416&viz=line_chart&years=2001,2019).
* **Adaptación educativa**: Los archivos han sido modificados intencionalmente para incorporar desafíos técnicos que permiten aplicar los contenidos abordados en clases, tales como limpieza de datos, normalización, detección de duplicados, y combinación de fuentes.


**Descripción de los archivos disponibles**

* **`libertad_prensa_codigo.csv`**: Contiene los pares `codigo_iso` y `pais`. Incluye intencionalmente un código ISO con dos nombres distintos de país para efectos de limpieza y validación de datos.

* **`libertad_prensa_01.csv`**: Contiene registros de los años **anteriores a 2010**. Incluye las variables `PAIS`, `ANIO`, `INDICE`, y `RANKING` con nombres de columna en **mayúsculas**.

* **`libertad_prensa_02.csv`**: Contiene registros de los años **desde 2010 en adelante**. Estructura similar al archivo anterior, con nombres de columna también en **mayúsculas**.





In [58]:
import numpy as np
import pandas as pd

# lectura de datos
archivos_anio = [
    'https://raw.githubusercontent.com/fralfaro/MAT306/main/docs/labs/data/libertad_prensa_01.csv',
    'https://raw.githubusercontent.com/fralfaro/MAT306/main/docs/labs/data/libertad_prensa_02.csv'
 ]
df_codigos = pd.read_csv('https://raw.githubusercontent.com/fralfaro/MAT306/main/docs/labs/data/libertad_prensa_codigo.csv')



### 1. Consolidación y limpieza de datos

A partir de los archivos disponibles, realice los siguientes pasos:

**a)** Cree un DataFrame llamado `df_anio` que consolide la información proveniente de los archivos **`libertad_prensa_01.csv`** y **`libertad_prensa_02.csv`**, correspondientes a distintas ventanas de tiempo. Recuerde que ambos archivos tienen nombres de columnas en mayúscula, por lo que debe normalizarlas a **minúscula** para asegurar consistencia.

**b)** Explore el archivo **`libertad_prensa_codigo.csv`** e identifique el código ISO que aparece asociado a dos nombres de país distintos. Elimine el registro que corresponda a un valor incorrecto o inconsistente, conservando solo el que considere válido.

**c)** Una vez preparados los archivos, cree un nuevo DataFrame llamado `df` que combine `df_anio` con `df_codigos`, utilizando la columna `codigo_iso` como clave. Asegúrese de realizar una unión que conserve únicamente los registros que tengan coincidencia en ambas fuentes.

> **Sugerencia**:
>
> * Para unir los archivos por filas (años), utilice la función `pd.concat([...])`.
> * Para combinar información por columnas (variables), utilice `pd.merge(...)` especificando `on='codigo_iso'`.



**Respuesta PARTE 1**

**a)** Cree un DataFrame llamado `df_anio` que consolide la información proveniente de los archivos **`libertad_prensa_01.csv`** y **`libertad_prensa_02.csv`**, correspondientes a distintas ventanas de tiempo. Recuerde que ambos archivos tienen nombres de columnas en mayúscula, por lo que debe normalizarlas a **minúscula** para asegurar consistencia.
> * Para unir los archivos por filas (años), utilice la función `pd.concat([...])`.

In [59]:
l_p_01 = pd.read_csv(archivos_anio[0], sep=",")
l_p_01.columns = l_p_01.columns.str.lower()
l_p_02 = pd.read_csv(archivos_anio[1], sep=",")
l_p_02.columns = l_p_02.columns.str.lower()

In [60]:
l_p_01.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1620 entries, 0 to 1619
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   codigo_iso  1620 non-null   object 
 1   anio        1620 non-null   int64  
 2   indice      1439 non-null   float64
 3   ranking     1439 non-null   float64
dtypes: float64(2), int64(1), object(1)
memory usage: 50.8+ KB


In [61]:
l_p_02.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1440 entries, 0 to 1439
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   codigo_iso  1440 non-null   object 
 1   anio        1440 non-null   int64  
 2   indice      1225 non-null   float64
 3   ranking     1398 non-null   float64
dtypes: float64(2), int64(1), object(1)
memory usage: 45.1+ KB


In [62]:
df_anio = pd.concat([l_p_01, l_p_02]).reset_index(drop=True)
df_anio

codigo_iso  anio  indice  ranking
0           AFG  2001   35.50     59.0
1           AGO  2001   30.20     50.0
2           ALB  2001     NaN      NaN
3           AND  2001     NaN      NaN
4           ARE  2001     NaN      NaN
...         ...   ...     ...      ...
3055        WSM  2019   18.25     22.0
3056        YEM  2019   61.66    168.0
3057        ZAF  2019   22.19     31.0
3058        ZMB  2019   36.38    119.0
3059        ZWE  2019   42.23    127.0

[3060 rows x 4 columns]

In [63]:
df_anio.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3060 entries, 0 to 3059
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   codigo_iso  3060 non-null   object 
 1   anio        3060 non-null   int64  
 2   indice      2664 non-null   float64
 3   ranking     2837 non-null   float64
dtypes: float64(2), int64(1), object(1)
memory usage: 95.8+ KB


**b)** Explore el archivo **`libertad_prensa_codigo.csv`** e identifique el código ISO que aparece asociado a dos nombres de país distintos. Elimine el registro que corresponda a un valor incorrecto o inconsistente, conservando solo el que considere válido.

In [64]:
df_codigos#.head()

codigo_iso                    pais
0          AFG             Afghanistán
1          AGO                  Angola
2          ALB                 Albania
3          AND                 Andorra
4          ARE  Emiratos Árabes Unidos
..         ...                     ...
176        YEM                   Yemen
177        ZAF               Sudáfrica
178        ZMB                  Zambia
179        ZWE                Zimbabue
180        ZWE                    malo

[181 rows x 2 columns]

In [65]:
df_codigos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 181 entries, 0 to 180
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   codigo_iso  181 non-null    object
 1   pais        181 non-null    object
dtypes: object(2)
memory usage: 3.0+ KB


In [66]:
df_codigos['codigo_iso'].duplicated().sum()

np.int64(1)

In [67]:
df_codigos.loc[df_codigos['codigo_iso'].duplicated()]

codigo_iso  pais
180        ZWE  malo

In [68]:
df_codigos.loc[df_codigos['codigo_iso']=='ZWE']

codigo_iso      pais
179        ZWE  Zimbabue
180        ZWE      malo

In [69]:
df_codigos.drop(index=180,inplace=True)
df_codigos.tail()

codigo_iso       pais
175        WSM      Samoa
176        YEM      Yemen
177        ZAF  Sudáfrica
178        ZMB     Zambia
179        ZWE   Zimbabue

Notemos que la situación inversa se da también.

In [72]:
df_codigos['pais'].duplicated().sum()

np.int64(1)

In [73]:
df_codigos.loc[df_codigos['pais'].duplicated()]

codigo_iso     pais
119        NGA  Nigeria

In [74]:
df_codigos.loc[df_codigos['pais']=='Nigeria']

codigo_iso     pais
118        NER  Nigeria
119        NGA  Nigeria

Dado que la instrucción **sólo** indica eliminar el dato cuyo ISO corresponde a dos nombres, se procede de este modo.

Con todo, el código que resolvería este nuevo problema se detalla en la próxima celda, comentado (bastaría descomentar la primera línea y volver a correr el código para eliminar el nuevo dato indeseado).

In [ ]:
#df_codigos.drop(index=118,inplace=True)
# el código ISO de Nigeria es NGA

**c)** Una vez preparados los archivos, cree un nuevo DataFrame llamado `df` que combine `df_anio` con `df_codigos`, utilizando la columna `codigo_iso` como clave. Asegúrese de realizar una unión que conserve únicamente los registros que tengan coincidencia en ambas fuentes.

> * Para combinar información por columnas (variables), utilice `pd.merge(...)` especificando `on='codigo_iso'`.

In [70]:
df = pd.merge(df_anio, df_codigos, on='codigo_iso', how='inner')
df

codigo_iso  anio  indice  ranking                    pais
0           AFG  2001   35.50     59.0             Afghanistán
1           AGO  2001   30.20     50.0                  Angola
2           ALB  2001     NaN      NaN                 Albania
3           AND  2001     NaN      NaN                 Andorra
4           ARE  2001     NaN      NaN  Emiratos Árabes Unidos
...         ...   ...     ...      ...                     ...
3055        WSM  2019   18.25     22.0                   Samoa
3056        YEM  2019   61.66    168.0                   Yemen
3057        ZAF  2019   22.19     31.0               Sudáfrica
3058        ZMB  2019   36.38    119.0                  Zambia
3059        ZWE  2019   42.23    127.0                Zimbabue

[3060 rows x 5 columns]



### 2. Exploración inicial del conjunto de datos

Una vez que hayas consolidado el DataFrame final `df`, realiza un análisis exploratorio básico respondiendo las siguientes preguntas:

#### **Estructura del DataFrame**

* ¿Cuántas **filas (observaciones)** contiene el conjunto de datos?
* ¿Cuántas **columnas** tiene el DataFrame?
* ¿Cuáles son los **nombres de las columnas**?
* ¿Qué **tipo de datos** tiene cada columna?
* ¿Hay columnas con un tipo de dato inesperado (por ejemplo, fechas como strings)?

#### **Resumen estadístico**

* Genera un resumen estadístico del conjunto de datos con `.describe()`.
  ¿Qué observas sobre los valores de `indice` y `ranking`?
* ¿Qué valores mínimo, máximo y promedio tiene la columna `indice`?
* ¿Qué países presentan los valores extremos en `indice` y `ranking`?

#### **Datos faltantes**

* ¿Cuántos valores nulos hay en cada columna?
* ¿Qué proporción de observaciones tienen valores faltantes?
* ¿Hay columnas con más del 30% de datos faltantes?

#### **Unicidad y duplicados**

* ¿Cuántos países distintos (`pais`) hay en el DataFrame?
* ¿Cuántos años distintos (`anio`) hay representados?
* ¿Existen filas duplicadas (exactamente iguales)? ¿Cuántas?

#### **Validación cruzada de columnas**

* ¿Hay inconsistencias entre el país (`pais`) y su código (`codigo_iso`)?
  (por ejemplo, un mismo código ISO asociado a más de un país)

> **Sugerencia**: Apoya tu análisis con funciones como `.info()`, `.nunique()`, `.isnull().sum()`, `.duplicated()`, `.value_counts()`, entre otras.



    

**Respuesta PARTE 2**

#### **Estructura del DataFrame**

* ¿Cuántas **filas (observaciones)** contiene el conjunto de datos?
* ¿Cuántas **columnas** tiene el DataFrame?
* ¿Cuáles son los **nombres de las columnas**?
* ¿Qué **tipo de datos** tiene cada columna?
* ¿Hay columnas con un tipo de dato inesperado (por ejemplo, fechas como strings)?

In [78]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3060 entries, 0 to 3059
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   codigo_iso  3060 non-null   object 
 1   anio        3060 non-null   int64  
 2   indice      2664 non-null   float64
 3   ranking     2837 non-null   float64
 4   pais        3060 non-null   object 
dtypes: float64(2), int64(1), object(2)
memory usage: 119.7+ KB


* El conjunto tiene 3060 observaciones.
* `df` tiene 5 columnas.
* Las columnas son: `codigo_iso`, `anio`, `indice`, `ranking` y `pais`, en ese orden.
* `anio` tiene datos del tipo `int64`, y tanto `indice` como `ranking` tienen datos del tipo `float64`, mientras que `codigo_iso` t `pais` (no numéricos) tiene datos tipo `object`.
* Es llamativo que ranking tenga datos tipo float, ya que, toda vez que indica un puesto en un listado, podría ser mejor representado por un entero.

#### **Resumen estadístico**

* Genera un resumen estadístico del conjunto de datos con `.describe()`.
  ¿Qué observas sobre los valores de `indice` y `ranking`?
* ¿Qué valores mínimo, máximo y promedio tiene la columna `indice`?
* ¿Qué países presentan los valores extremos en `indice` y `ranking`?

In [79]:
df.describe()

anio        indice        ranking
count  3060.000000   2664.000000    2837.000000
mean   2009.941176    205.782316     477.930913
std       5.786024   2695.525264    6474.935347
min    2001.000000      0.000000       1.000000
25%    2005.000000     15.295000      34.000000
50%    2009.000000     28.000000      70.000000
75%    2015.000000     41.227500     110.000000
max    2019.000000  64536.000000  121056.000000

* En ambos el salto desde el 3er cuartil al máximo es muy brusco: tienen outliers en la parte superior de su rango.
* `indice` tiene un mínimo de 0, un máximo de 64536 y un promedio de, aproximadamente, 205.78.
* Abajo se detallan los países que han presentado los valores extremos de `indice` y `ranking` en alguno de los años considerados en las observaciones.

In [83]:
df.loc[df['indice'] == 0]['pais'].unique()

array(['Dinamarca', 'Finlandia', 'Irlanda', 'Noruega', 'Suecia', 'Suiza',
       'Islandia', 'Países Bajos'], dtype=object)

In [84]:
df.loc[df['ranking'] == 1]['pais'].unique()

array(['Finlandia', 'Islandia', 'Países Bajos', 'Noruega', 'Suiza',
       'Dinamarca', 'Irlanda', 'Eslovaquia', 'Luxemburgo', 'Estonia',
       'Austria', 'Nueva Zelanda'], dtype=object)

In [85]:
df.loc[df['indice'] == 64536]['pais'].unique()

array(['Kosovo'], dtype=object)

In [86]:
df.loc[df['ranking'] == 121056]['pais'].unique()

array(['Kosovo'], dtype=object)

#### **Datos faltantes**

* ¿Cuántos valores nulos hay en cada columna?
* ¿Qué proporción de observaciones tienen valores faltantes?
* ¿Hay columnas con más del 30% de datos faltantes?

In [89]:
df.isna().sum()

codigo_iso      0
anio            0
indice        396
ranking       223
pais            0
dtype: int64

* Sólo `indice` y `ranking` tienen valores nulos, con un total de 396 y 223 tales valores, respectivamente.

In [104]:
df.loc[(df['indice'].isna()) + (df['ranking'].isna())]
# '+' obra como 'or' (1 y 2 son True, 0 es False)

codigo_iso  anio  indice  ranking                    pais
2           ALB  2001     NaN      NaN                 Albania
3           AND  2001     NaN      NaN                 Andorra
4           ARE  2001     NaN      NaN  Emiratos Árabes Unidos
6           ARM  2001     NaN      NaN                 Armenia
7           ATG  2001     NaN      NaN       Antigua y Barbuda
...         ...   ...     ...      ...                     ...
2789        KSV  2018     NaN      NaN                  Kosovo
2887        ATG  2019     NaN      NaN       Antigua y Barbuda
2894        BFA  2019   24.53      NaN            Burkina Faso
2945        GRD  2019     NaN      NaN                 Granada
2969        KSV  2019     NaN      NaN                  Kosovo

[397 rows x 5 columns]

In [105]:
len(df.loc[(df['indice'].isna()) + (df['ranking'].isna())])

397

In [106]:
len(df.loc[(df['indice'].isna()) + (df['ranking'].isna())]) / df.shape[0]

0.12973856209150328

* Aproximadamente un 12.97% de las observaciones tiene algún dato nulo.

In [95]:
df.isna().sum()/df.shape[0]*100

codigo_iso     0.000000
anio           0.000000
indice        12.941176
ranking        7.287582
pais           0.000000
dtype: float64

* No, no hay columnas con más de un 30% de datos nulos.

#### **Unicidad y duplicados**

* ¿Cuántos países distintos (`pais`) hay en el DataFrame?
* ¿Cuántos años distintos (`anio`) hay representados?
* ¿Existen filas duplicadas (exactamente iguales)? ¿Cuántas?

In [107]:
df['pais'].nunique()

179

* Existen 179 países distintos en `df`.

In [108]:
df['anio'].nunique()

17

* Hay 17 años representados en `df`.

In [112]:
df.duplicated().sum()

np.int64(0)

* No existen filas duplicadas en `df`.

#### **Validación cruzada de columnas**

* ¿Hay inconsistencias entre el país (`pais`) y su código (`codigo_iso`)?
  (por ejemplo, un mismo código ISO asociado a más de un país)

Como sabemos, sólo un código ISO estaba asociado a más de un país, y tal situación fue resuelta; mientras que, al revés, Nigeria sigue teniendo dos códigos ISO asociados, lo cual se traduce en que las entradas correspondientes se encuentran duplicadas en `df`.

In [77]:
df.loc[df['pais']=='Nigeria'].head()

codigo_iso  anio  indice  ranking     pais
118        NER  2001   18.50     17.0  Nigeria
119        NGA  2001   15.50     14.0  Nigeria
298        NER  2002   15.75     16.0  Nigeria
299        NGA  2002   31.50     55.0  Nigeria
478        NER  2003   18.33     26.0  Nigeria




### 3. Comparación regional: países latinoamericanos

En esta sección se busca identificar cuáles son los países de América Latina que han presentado los valores extremos del **Índice de Libertad de Prensa** en cada año observado.

> Recuerda que un menor puntaje en `indice` implica mayor libertad de prensa.

#### **Tareas:**

**a)** Utilizando un ciclo `for`, recorre cada año del conjunto de datos filtrado por países latinoamericanos, y determina para cada año:

* El país con el menor valor de `indice` (mayor libertad de prensa).
* El país con el mayor valor de `indice` (menor libertad de prensa).

**b)** Resuelve la misma tarea del punto anterior utilizando un enfoque vectorizado con `groupby`, sin usar ciclos explícitos.



#### **Lista de países latinoamericanos considerada:**

```python
america = ['ARG', 'ATG', 'BLZ', 'BOL', 'BRA', 'CAN', 'CHL', 'COL', 'CRI',
           'CUB', 'DOM', 'ECU', 'GRD', 'GTM', 'GUY', 'HND', 'HTI', 'JAM',
           'MEX', 'NIC', 'PAN', 'PER', 'PRY', 'SLV', 'SUR', 'TTO', 'URY',
           'USA', 'VEN']
```

> Puedes usar esta lista para filtrar el DataFrame final por la columna `codigo_iso`.



In [ ]:
# respuesta
america = ['ARG', 'ATG', 'BLZ', 'BOL', 'BRA', 'CAN', 'CHL', 'COL', 'CRI',
       'CUB', 'DOM', 'ECU', 'GRD', 'GTM', 'GUY', 'HND', 'HTI', 'JAM',
       'MEX', 'NIC', 'PAN', 'PER', 'PRY', 'SLV', 'SUR', 'TTO', 'URY',
       'USA', 'VEN']

df_america =  pd.DataFrame() # FIX ME

### 4. Análisis anual del índice por país

En esta sección se busca analizar la evolución del **índice máximo** de libertad de prensa alcanzado por cada país a lo largo del tiempo.

#### **Tarea principal:**

* Construye una tabla dinámica (`pivot_table`) donde las **filas** correspondan a los países, las **columnas** a los años (`anio`) y los **valores** sean el `indice` máximo alcanzado por cada país en ese año.
* Asegúrate de reemplazar los valores nulos resultantes con `0`.

> **Hint**: Puedes utilizar el parámetro `fill_value=0` en `pd.pivot_table(...)`.



#### **Preguntas adicionales:**

**a)** ¿Qué país tiene el mayor valor de `indice` en toda la tabla resultante? ¿Y cuál tiene el menor (distinto de cero)?
**b)** ¿Qué años presentan en promedio los valores de `indice` más altos? ¿Y los más bajos?

> (Pista: usa `.mean(axis=0)` sobre la tabla pivot)

**c)** ¿Qué país muestra mayor **variabilidad** (diferencia entre su máximo y mínimo `indice` a lo largo del tiempo)?

> (Pista: aplica `.max(axis=1) - .min(axis=1)`)

**d)** ¿Existen países con índice constante a lo largo de todos los años registrados? ¿Cuáles?

**e)** ¿Qué países no tienen ningún dato (es decir, quedaron con todos los valores igual a 0)? ¿Podrías explicar por qué?





In [ ]:
# FIX ME